### PR TITLE
add plot for errors over different lead times

### PR DESCRIPTION
**Pull request by sdtaylor**
also a function to calculate those errors

### CODE REVIEW

File: forecast_tools.R [hunk 1]
```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
...
+#' Error can be any function. The level, species, and currency columns from
+#' observations and forecasts must have matching values.
+#' 
+#' Will only return values where there are matching comparison columns (currency, level, species)
+#' 
+#' @param observations dataframe Has the columns NewMoonNumber, currency, level, species, actual
+#' @param forecasts dataframe passes the forecast validity check. Must have matching values in
+#'                  the comparison columns
+#' @param error_metric chr either 'mse' for mean squared error or 'likelihood' for the likelihood of the observation
+#' assumring the estimate and PI's fit a normal distribution
```
Comment (ethanwhite) [inline forecast_tools.R]: `assumring` -> `assuming`

File: forecast_tools.R [hunk 2]
```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
...
+#' @param ci_value int The value of the forecast confidence interval to scale PI values for the likelihood metric
```
Comment (ethanwhite) [inline forecast_tools.R]: I don't see this parameter getting used and am unclear on exactly what it is based on the description.

File: forecast_tools.R [hunk 3]
```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
...
+#' @return data.frame Data.frame with the columns model, error, lead_time, level, species, currency
+#' 
+calculate_forecast_error = function(observations, forecasts, error_metric='MSE', ci_value=90){
+  #The tibble datatype output from dplyr causes issues here
+  observations = as.data.frame(observations)
+  forecasts = as.data.frame(forecasts)
+  
+  if(!forecast_is_valid(forecasts)) stop('Forecast datafarme not valid')
```
Comment (ethanwhite) [inline forecast_tools.R]: `datafarme` -> `data frame`

File: forecast_tools.R [hunk 4]
```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
...
+  
+  valid_observation_columns=c('NewMoonNumber','currency','level','species','actual')
```
Comment (ethanwhite) [inline forecast_tools.R]: Add spaces around `=`

File: forecast_tools.R (hunk 2 above)
Comment (ethanwhite) [reply forecast_tools.R]: Ah, I see, this is going to get implemented when the likelihood option is added.

File: forecast_tools.R [hunk 5]
```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
...
+    group_by(date) %>%
+    summarise(new_moon_of_forecast = min(NewMoonNumber)-1) %>%
+    ungroup()
+  
+  comparison = comparison %>%
+    left_join(forecast_date_new_moon_number, by='date') %>%
+    mutate(lead_time=NewMoonNumber - new_moon_of_forecast) %>%
+    select(-new_moon_of_forecast, -NewMoonNumber, -date)
+  
+  comparison = comparison %>%
```
Comment (ethanwhite) [inline forecast_tools.R]: Any reason not to have one continuous pipe here? If it's more readable split up I think having a separately named intermediate variable would make the logic easier to follow.

File: forecast_tools.R [hunk 6]
```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
...
+#' 
+#' @param error_df data.frame The output from calculate_foreast_error()
+#' @param level str Valid level
+#' @param species str Valid species
+#' @param currency str Valid currency
+#' @param error_metric str error metric used
+plot_lead_time_errors=function(error_df, level, species, currency, error_metric){
+  plot_title = paste0('Level: ',level,', Species: ',species,', Currency: ',currency)
+  
+  graph=ggplot(error_df, aes(x=lead_time, y=error, group=model, color=model)) +
```
Comment (ethanwhite) [inline forecast_tools.R]: Add spaces around `=`

File: viz_maincode.R [hunk 7]
```diff
@@ -15,3 +15,23 @@ plot_sp_predicts(ensemble, 'All')
 
 # Plot forecasts and outcomes
 plot_sp_predicts(ensemble, 'All', observed = obs_data)
+
+new_moons=read.csv('~/PortalData/Rodents/moon_dates.csv')
+
+#Format raw observations ot match forecast comparison columns and include NewMoonNumbers
```
Comment (ethanwhite) [inline viz_maincode.R]: `ot` -> `to`

**Review by ethanwhite — COMMENTED** [review]
Looks great overall. I've made a few suggestions for minor cleanup.

**Comment by sdtaylor** [comment]
Thanks for finding my typos. For the pipe I generally just try to keep them short. How's this for descriptive names?

**Comment by ethanwhite** [comment]
> For the pipe I generally just try to keep them short. How's this for descriptive names?

Yeah, I think that's a good way to go. The naming looks great.

**Pull request merged**